File: server_integration_tests/kotlin/kotlin_client/client-app/src/test/kotlin/dev/graphlink/kotlinclienttest/QueriesTest.kt
package dev.graphlink.kotlinclienttest

import dev.graphlink.kotlinclient.generated.client.GraphLinkClient
import kotlinx.coroutines.test.runTest
import kotlin.test.BeforeTest
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNotNull
import kotlin.test.assertNull
import kotlin.test.assertTrue

class QueriesTest {

    private lateinit var client: GraphLinkClient

    @BeforeTest
    fun setUp() {
        client = Fixtures.newClient()
    }

    @Test
    fun getAuthor_returnsAuthorById() = runTest {
        val res = client.queries.getAuthor("1")
        assertNotNull(res.getAuthor)
        assertEquals("1", res.getAuthor!!.id)
        assertEquals("Ramdane", res.getAuthor!!.name)
    }

    @Test
    fun getAuthor_returnsNullForUnknownId() = runTest {
        val res = client.queries.getAuthor("missing")
        assertNull(res.getAuthor)
    }

    @Test
    fun getAuthor_articlesResolvedViaBatchMapping() = runTest {
        val res = client.queries.getAuthor("1")
        val titles = res.getAuthor!!.articles?.map { it.title }?.toSet()
        assertTrue(titles?.containsAll(setOf("GraphLink Basics", "Advanced GraphLink")) ?: false)
    }

    @Test
    fun getArticle_returnsArticleById() = runTest {
        val res = client.queries.getArticle("1")
        assertEquals("1", res.getArticle.id)
        assertEquals("GraphLink Basics", res.getArticle.title)
    }

    @Test
    fun getArticle_authorResolvedViaSchemaMapping() = runTest {
        val res = client.queries.getArticle("1")
        assertEquals("1", res.getArticle.author?.id)
        assertEquals("Ramdane", res.getArticle.author?.name)
    }

    @Test
    fun listAuthors_returnsAllAuthors() = runTest {
        val res = client.queries.listAuthors()
        assertEquals(setOf("1", "2"), res.listAuthors.map { it.id }.toSet())
    }

    @Test
    fun listArticles_returnsAllArticles() = runTest {
        val res = client.queries.listArticles()
        assert(res.listArticles.any { it.id == "1" })
    }
}

File: server_integration_tests/clients/swift_client/Tests/GraphLinkGeneratedTests/QueriesTest.swift
import XCTest
@testable import GraphLinkGenerated

final class QueriesTest: XCTestCase {
    var client: GraphLinkClient!

    override func setUp() {
        client = Fixtures.newClient()
    }

    func testGetAuthor_returnsAuthorById() async throws {
        let res = try await client.queries.getAuthor(id: "1", fieldArgs: GetAuthorFieldArgs(latestArticlesLimit: 10))
        XCTAssertNotNil(res.getAuthor)
        XCTAssertEqual(res.getAuthor!.id, "1")
        XCTAssertEqual(res.getAuthor!.name, "Ramdane")
    }

    func testGetAuthor_returnsNullForUnknownId() async throws {
        let res = try await client.queries.getAuthor(id: "missing", fieldArgs: GetAuthorFieldArgs(latestArticlesLimit: 10))
        XCTAssertNil(res.getAuthor)
    }

    func testGetAuthor_articlesResolvedViaBatchDataLoaderMapping() async throws {
        let res = try await client.queries.getAuthor(id: "1", fieldArgs: GetAuthorFieldArgs(latestArticlesLimit: 10))
        let titles = Set(res.getAuthor!.articles!.map { $0.title })
        XCTAssertTrue(titles.isSuperset(of: ["GraphLink Basics", "Advanced GraphLink"]))
    }

    func testGetAuthor_latestArticlesResolvedViaNonBatchMappingWithArguments() async throws {
        let res = try await client.queries.getAuthor(id: "1", fieldArgs: GetAuthorFieldArgs(latestArticlesLimit: 10))
        let titles = Set(res.getAuthor!.latestArticles!.map { $0.title })
        XCTAssertTrue(titles.isSuperset(of: ["GraphLink Basics", "Advanced GraphLink"]))
    }

    func testGetAuthorWithoutArticle_doesNotRequireLimitArgumentWhenArticlesNotProjected() async throws {
        let res = try await client.queries.getAuthorWithoutArticle(id: "1")
        XCTAssertNotNil(res.getAuthor)
        XCTAssertEqual(res.getAuthor!.id, "1")
        XCTAssertEqual(res.getAuthor!.name, "Ramdane")
    }

    func testGetArticle_returnsArticleById() async throws {
        let res = try await client.queries.getArticle(id: "1", fieldArgs: GetArticleFieldArgs(latestArticlesLimit: 2))
        XCTAssertEqual(res.getArticle.id, "1")
        XCTAssertEqual(res.getArticle.title, "GraphLink Basics")
    }

    func testGetArticle_authorResolvedViaNonBatchSchemaMapping() async throws {
        let res = try await client.queries.getArticle(id: "1", fieldArgs: GetArticleFieldArgs(latestArticlesLimit: 2))
        XCTAssertEqual(res.getArticle.author?.id, "1")
        XCTAssertEqual(res.getArticle.author?.name, "Ramdane")
    }

    func testGetAuthorAndArticle_resolvesBothRootFieldsWithOwnArgsPlusSharedFragmentArgument() async throws {
        let res = try await client.queries.getAuthorAndArticle(
            authorId: "1", articleId: "3", fieldArgs: GetAuthorAndArticleFieldArgs(latestArticlesLimit: 1)
        )
        XCTAssertEqual(res.author!.id, "1")
        XCTAssertEqual(res.author!.latestArticles!.count, 1)
        XCTAssertEqual(res.article.id, "3")
        XCTAssertEqual(res.article.title, "Alice's First Post")
    }

    func testListAuthors_listAuthorsReturnsAllAuthors() async throws {
        let res = try await client.queries.listAuthors(fieldArgs: ListAuthorsFieldArgs(latestArticlesLimit: 10))
        let ids = Set(res.listAuthors.map { $0.id })
        XCTAssertTrue(ids.isSuperset(of: ["1", "2"]))
    }

    func testListArticles_listArticlesReturnsAllArticles() async throws {
        let res = try await client.queries.listArticles(fieldArgs: ListArticlesFieldArgs(latestArticlesLimit: 2))
        XCTAssertTrue(res.listArticles.map { $0.id }.contains("1"))
    }
}
